constants.go

```diff
@@ -577,7 +580,14 @@ const (
 const OSSMigratedV6 = "migrate-v6.0"
 
 // MinClientVersion is the minimum client version required by the server.
-const MinClientVersion = "3.0.0"
+var MinClientVersion string
+
+func init() {
+	// Per https://github.com/gravitational/teleport/blob/master/rfd/0012-teleport-versioning.md,
+	// only one major version backwards is supported for clients.
+	ver := semver.New(Version)
+	MinClientVersion = fmt.Sprintf("%d.0.0", ver.Major-1)
+}
```

Comment:
@awly What are the implications of this? Future version of `tsh` will return an error if talking a new server? For example `tsh` 6.0 will return an error when talking to a `8.0` server?

Reply:
Yes, by default `tsh` 6.0 will return an error when hitting the `Ping` endpoint on a `8.0` proxy.
There's a global `--skip-version-check` flag in `tsh` to bypass this.

Reply:
My understanding of that RFD was that is what we would guarantee to work (test and support) but not move to actively break things that work.

For example, `tsh` 3.1 continues to work with Teleport 6.0. Why do we want to introduce this breaking change, what problem is being solved?

Reply:
To prevent people from hitting weird issues due to using outdated clients.

I don't feel particularly strong about this, just noticed that we provide this field without ever actually updating it. Maybe it should be totally removed, along with the attached logic?

Reply:
I like the idea of actually updating `MinClientVersion` in this way, this will also be useful for the API's version check. 

Returning an error in tsh might not be great. I would instead log a warning that says something like "WARNING: server version x is not guarenteed to be compatible with server version y".

You could also wrap the error with "version check can be skipped with --skip-version-check flag" to give users a clear path to avoiding the error. This could be a nice way to prod users to upgrade without actually denying functionality.

Reply:
If users attempt to use a 3.1 `tsh` client with an 8.0 `teleport` server it might appear to work at first, but there are bound to be subtle bugs and implementation issues which might send people down a rabbit hole unnecessarily and unnecessarily waste a lot of everyone's time. We have no provisions at all in our test plan for deliberately using old versions of `tsh`.

Furthermore, customers are likely to be missing out on useful bug fixes, improvements and new features if they don't proactively upgrade their `tsh` client occasionally. This is the same reason why MacOS, Windows, iOS, Android and even Linux all have automatic software updates built in and use various different types of carrots and sticks to encourage or even force users into installing them.

I like the idea of us at least warning users that their `tsh` client version is out of date with the server and could/should be updated. Even better if we work on implementing a `tsh upgrade` feature in future so they can easily install a newer version which is appropriate for the server they're connecting to.

Reply:
@russjones WDYT about changing the `tsh` version check logic to only print a warning and keeping this dynamic version updated?

If not, we should at least add a step in the testplan to verify very old clients and keep `MinClientVersion` manually updated.

Reply:
I like @Joerger's idea, but let's address after 6.1 since I don't think it should just be `log.Warnf`.

We have a lot of functionality to merge in within the next week.

Reply:
@russjones @Joerger turns out, all the necessary logic and CLI flag are already in place.

I changed `teleport.Version` to and older version and here's how it looks:
```
$ tsh version
Teleport v4.0.0-dev git:v6.0.0-alpha.2-396-g20e428ed7 go1.16.3

$ teleport version
Teleport v7.0.0-dev git:v6.0.0-alpha.2-396-g20e428ed7 go1.16.3

$ tsh login --proxy=localhost --insecure
ERROR: minimum client version supported by the server is 6.0.0. Please upgrade the client, downgrade the server, or use the --skip-version-check flag to by-pass this check.

$ tsh login --proxy=localhost --insecure --skip-version-check
Enter password for Teleport user awly:
WARNING: You are using insecure connection to SSH proxy https://localhost:3080
> Profile URL:        https://localhost:3080
  Logged in as:       awly
  Cluster:            talos
  Roles:              admin
  Logins:             awly
  Kubernetes:         disabled
  Valid until:        2021-04-17 02:27:53 -0700 PDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
```

Is this good enough @russjones ?

Reply:
@awly What do you think about making the error a warning and telling people that it may break in weird ways and to upgrade or use the suppression flag. This way no functionality is broken but we can send the message we want to users (please upgrade), other than that looks good to me.

Reply:
Done.
The output looks like this now:
```sh
$ tsh login --proxy=localhost --insecure
WARNING: detected potentially incompatible client and server versions: minimum client version supported by the server is 6.0.0 but you are using 5.0.0-dev. Please upgrade the client, downgrade the server, or use the --skip-version-check flag to by-pass this check.
         Future versions of tsh will fail when incompatible versions are detected.
Enter password for Teleport user awly:
WARNING: You are using insecure connection to SSH proxy https://localhost:3080
> Profile URL:        https://localhost:3080
  Logged in as:       awly
  Cluster:            talos
  Roles:              admin
  Logins:             awly
  Kubernetes:         disabled
  Valid until:        2021-05-07 01:48:05 -0700 PDT [valid for 12h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
```

Reply:
*bypass

Looks great otherwise.